xla/tests/collective_ops_e2e_test.cc

```diff
@@ -831,6 +866,48 @@ ENTRY main.9_spmd {
   CollectiveOpsCompareWindowedNonWindowed(kModuleReplicatedStr);
 }
 
+TEST_F(CollectiveOpsTestE2EWindowedNonWindowed,
+       WindowedEinsumE2EAllgatherMultiConsumerF8) {
+  absl::string_view kModuleReplicatedStr = R"(
+HloModule pjit__unnamed_wrapped_function_, entry_computation_layout={(f8e4m3fn[2,16,48]{2,1,0}, f8e4m3fn[48,192]{1,0}, f8e4m3fn[48,192]{1,0}, bf16[], bf16[], bf16[])->bf16[2,16,192]{2,1,0}}, allow_spmd_sharding_propagation_to_parameters={false,false,false,false}, num_partitions=4
+
+ENTRY main.12 {
+  Arg_0.1 = f8e4m3fn[2,16,48]{2,1,0} parameter(0), sharding={devices=[1,4,1]<=[4]}
+  Arg_1.2 = f8e4m3fn[48,192]{1,0} parameter(1), sharding={devices=[1,4]<=[4]}
+  Arg_2.3 = bf16[] parameter(3)
+  Arg_3.4 = bf16[] parameter(4)
+  broadcast = bf16[2,16,48]{2,1,0} broadcast(Arg_2.3), dimensions={}
+  broadcast.1 = bf16[48,192]{1,0} broadcast(Arg_3.4), dimensions={}
+  convert = bf16[2,16,48]{2,1,0} convert(Arg_0.1)
+  convert.1 = bf16[48,192]{1,0} convert(Arg_1.2)
+  multiply = bf16[2,16,48]{2,1,0} multiply(broadcast, convert)
+  multiply.1 = bf16[48,192]{1,0} multiply(broadcast.1, convert.1)
+  dot.5 = bf16[2,16,192]{2,1,0} dot(multiply, multiply.1), lhs_contracting_dims={2}, rhs_contracting_dims={0}
+  custom-call.7 = bf16[2,16,192]{2,1,0} custom-call(dot.5), custom_call_target="Sharding", sharding={devices=[1,1,4]<=[4]}
+  Arg_5.6 = f8e4m3fn[48,192]{1,0} parameter(2), sharding={devices=[1,4]<=[4]}
+  Arg_6.7 = bf16[] parameter(5)
+  broadcast.5 = bf16[48,192]{1,0} broadcast(Arg_6.7), dimensions={}
+  convert.3 = bf16[48,192]{1,0} convert(Arg_5.6)
+  multiply.2 = bf16[48,192]{1,0} multiply(broadcast.5, convert.3)
+  dot.6 = bf16[2,16,192]{2,1,0} dot(multiply, multiply.2), lhs_contracting_dims={2}, rhs_contracting_dims={0}
+  ROOT add.8 = bf16[2,16,192]{2,1,0} add(custom-call.7, dot.6)
+} // main.12
+)";
+
+  CollectiveOpsCompareWindowedNonWindowed(kModuleReplicatedStr,
+                                          /*disable_dot_merger=*/true);
+
+  // Verify the creation of FP8 GEMM Custom Calls on Hopper and newer
+  // architectures.
+  DebugOptions opts = GetDebugOptionsForTest();
+  opts.set_xla_gpu_threshold_for_windowed_einsum_mib(0);
+  opts.set_xla_gpu_multi_streamed_windowed_einsum(true);
+  opts.set_xla_gpu_graph_min_graph_size(200);
+  opts.set_xla_gpu_enable_triton_gemm(false);
+  opts.add_xla_disable_hlo_passes("dot-merger");
+  CollectiveOpsVerifyF8Matmul(kModuleReplicatedStr, opts);
```

Comment:
Why are these options needed to test the windowed einsum handler? 
What does this have to do with the dot-merger? Why do you have to disable triton?

Reply:
The purpose of this end-to-end test is to verify the sequence of windowed einsum handler and GEMM rewriter passes. The dot merger and Triton passes are disabled since they can potentially interfere with the outcome.

Reply:
Do you know why they interfer with the outcome? Is there a fundamental reason or can this be implemented more generically so that it does compose well with the rest of XLA?